--- Day 6: Trash Compactor ---
After helping the Elves in the kitchen, you were taking a break and helping them re-enact a movie scene when you over-enthusiastically jumped into the garbage chute!

A brief fall later, you find yourself in a garbage smasher. Unfortunately, the door's been magnetically sealed.

As you try to find a way out, you are approached by a family of cephalopods! They're pretty sure they can get the door open, but it will take some time. While you wait, they're curious if you can help the youngest cephalopod with her math homework.

Cephalopod math doesn't look that different from normal math. The math worksheet (your puzzle input) consists of a list of problems; each problem has a group of numbers that need to be either added (+) or multiplied (*) together.

However, the problems are arranged a little strangely; they seem to be presented next to each other in a very long horizontal list. For example:

123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
Each problem's numbers are arranged vertically; at the bottom of the problem is the symbol for the operation that needs to be performed. Problems are separated by a full column of only spaces. The left/right alignment of numbers within each problem can be ignored.

So, this worksheet contains four problems:

123 * 45 * 6 = 33210
328 + 64 + 98 = 490
51 * 387 * 215 = 4243455
64 + 23 + 314 = 401
To check their work, cephalopod students are given the grand total of adding together all of the answers to the individual problems. In this worksheet, the grand total is 33210 + 490 + 4243455 + 401 = 4277556.

Of course, the actual worksheet is much wider. You'll need to make sure to unroll it completely so that you can read the problems clearly.

Solve the problems on the math worksheet. What is the grand total found by adding together all of the answers to the individual problems?

In [ ]:
with open('input.txt') as file:
    # transpose the matrix: Each line of input.txt should be split into elements with the elements appended sequentially to a list of lists
    first_line = next(file)
    math_problems = [[] for _ in range(len(first_line.split()))]
    file.seek(0)
    for line in file:
        for index, element in enumerate(line.split()):
            math_problems[index].append(element)

In [32]:
sum_total = 0
for *numbers, operator in math_problems:
    numbers = [int(number) for number in numbers]
    match operator:
        case '+':
             sum_total += sum(numbers)
        case '*':
            from math import prod
            sum_total += prod(numbers)
sum_total

6171290547579

Your puzzle answer was 6171290547579.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
The big cephalopods come back to check on how things are going. When they see that your grand total doesn't match the one expected by the worksheet, they realize they forgot to explain how to read cephalopod math.

Cephalopod math is written right-to-left in columns. Each number is given in its own column, with the most significant digit at the top and the least significant digit at the bottom. (Problems are still separated with a column consisting only of spaces, and the symbol at the bottom of the problem is still the operator to use.)

Here's the example worksheet again:

123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
Reading the problems right-to-left one column at a time, the problems are now quite different:

The rightmost problem is 4 + 431 + 623 = 1058
The second problem from the right is 175 * 581 * 32 = 3253600
The third problem from the right is 8 + 248 + 369 = 625
Finally, the leftmost problem is 356 * 24 * 1 = 8544
Now, the grand total is 1058 + 3253600 + 625 + 8544 = 3263827.

Solve the problems on the math worksheet again. What is the grand total found by adding together all of the answers to the individual problems?

In [49]:
'''
Notes: since numbers are not vertically aligned in the problem set in a consistent manner and alignment is fundamentally important in this variant, we need to rebuild the algorithm to track padding accurately. 

I think it is important to approach this problem one column at a time. We scan all lines in parallel one space at a time (starting from the right since that is how the problem was defined for us, although i don't think this matters since no lines start or end with whitespace). If the whole column is whitespace, we move on. Otherwise, we record the number top to bottom
'''
with open('input.txt') as file:
    data = [line[::-1] for line in file]

In [ ]:
from math import prod

num_digits = len(data) - 1
operators = data[-1]
sum_total = 0
numbers = []
for col, _ in enumerate(data[0]):
    if all(data[row][col].isspace() for row in range(num_digits + 1)):
        match operator:
            case '+':
                sum_total += sum(numbers)
            case '*':
                sum_total += prod(numbers)
        numbers = []
        continue
    if any(data[row][col].isdigit() for row in range(num_digits)):
        number_str = ""
        for row in range(num_digits):
            number_str += data[row][col]
        numbers.append(int(number_str))
    if not operators[col].isspace():
        operator = operators[col]

if numbers:
    match operator:
        case '+':
            sum_total += sum(numbers)
        case '*':
            sum_total += prod(numbers)

sum_total

8811937976367

That's the right answer! You are one gold star closer to decorating the North Pole.

You have completed Day 6! You can [Share] this victory or [Return to Your Advent Calendar].